### ISSUE TITLE
vars does not allow lists of dicts as argument (undocumented ansible feature)

### ISSUE THREAD

**Issue by Vanav** [ansible/schemas#401]
File: `roles/nginx-env/tasks/main.yml`:
```
- name: Run as root
  become: yes
  block:
  - name: Multiple domains and certificates
    include_role:
      name: nginx-env
    vars:
    - nginx_domains: '{{ nginx_role_loop_item.nginx_domains }}'
    - env_name: '{{ nginx_role_loop_item.env_name }}'
    - nginx_env_server_extra: '{{ nginx_role_loop_item.nginx_env_server_extra }}'
    loop: '{{ nginx_domains_extra | d([]) }}'
    loop_control:
      loop_var: nginx_role_loop_item
```
`ansible-lint` shows error:
```
schema: {'name': 'Multiple domains and certificates', 'include_role': {'name': 'nginx-env'}, 'vars': [{'nginx_domains': '{{ nginx_role_loop_item.nginx_domains }}'}, {'env_name': '{{ nginx_role_loop_item.env_name }}'}, {'nginx_env_server_extra': '{{ nginx_role_loop_item.nginx_env_server_extra }}'}], 'loop': '{{ nginx_domains_extra | d([]) }}', 'loop_control': {'loop_var': 'nginx_role_loop_item'}}
is not valid under any of the given schemas (schema[tasks])
roles/nginx-env/tasks/main.yml:1  Returned errors will not include exact line numbers, but they will mention
the schema name being used as a tag, like ``schema``,
``schema``.

This rule is not skippable and stops further processing of the file.

Schema bugs should be reported towards (https://github.com/ansible/schemas) project instead of ansible-lint.

If incorrect schema was picked, you might want to either:

* move the file to standard location, so its file is detected correctly.
* use ``kinds:`` option in linter config to help it pick correct file type.

You can skip specific rules or tags by adding them to your configuration file:
# .config/ansible-lint.yml
warn_list:  # or 'skip_list' to silence them completely
  - experimental  # all rules tagged as experimental
```

**Comment by ssbarnea** [comment]
Apparently you are not aware that blocks do not allow looping, ansible will give a syntax error with your example. Can you please update it to make it pass ansible own syntax check?

**Comment by ssbarnea** [comment]
If you use `ansible.builtin.include_role` it will work.

**Comment by Vanav** [comment]
Ansible doesn't complain about syntax and code is working as expected.
`ansible.builtin.include_role` makes no difference:

> schema: {'name': 'Multiple domains and certificates', 'ansible.builtin.include_role': {'name': 'nginx-env'}, 'vars': [{'nginx_domains': '{{ nginx_role_loop_item.nginx_domains }}'}, {'env_name': '{{ nginx_role_loop_item.env_name }}'}, {'nginx_env_server_extra': '{{ nginx_role_loop_item.nginx_env_server_extra }}'}], 'loop': '{{ nginx_domains_extra | d([]) }}', 'loop_control': {'loop_var': 'nginx_role_loop_item'}} is not valid under any of the given schemas (schema[tasks])

**Comment by ssbarnea** [comment]
The title is wrong because the schema failure has nothing to do with include_role and instead if caused by passing a **list of dictionaries to `vars:`**, something that is not documented as supported by ansible:

```
# undocumented way to declare vars
vars:
  - foo: var

# official way to declare vars:
vars:
  foo: bar
```

The big question is if we should allow the undocumented syntax to be used or not. If that is not accidental, we need to have ansible official documentation updated to mention this as allowed at https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#defining-variables-in-a-play

**Comment by Vanav** [comment]
Thank you, I've fixed syntax, now there are no errors. I'm sorry for misleading, error message was not detailed enough.

**Comment by ssbarnea** [comment]
@Vanav No problem. That is caused by the validator which is not able to give the best-possible error messages. I had to remove some keys until I spotted the root cause.

I will keep this bug open because it is still a problem and we will need some actions on it on our side. As you probably seen, we already got a bug report in core about it. If it is supported, it must be documented.

**Comment by ssbarnea** [comment]
This feature is deprecated since https://github.com/ansible/ansible/pull/78927 merged. So there is no real need to update the schema.
